### PR TITLE
OS-8231

### PR DESCRIPTION
**Pull request by danmcd**
Various piadm(1M) changes to generate `pi.rc` (see https://github.com/joyent/illumos-joyent/pull/372).

### CODE REVIEW

**Comment by danmcd** [comment]
Sample generated `pi.rc` file with 7 total PIs.  Default + 6 means it's two pages worth (out of a high of 3).

```
\
\ Generated by piadm(1M).
\

\ Assume mainmenu_options=4 for now.

set mainmenu_caption[5]="Alternate [P]latform Images..."
set mainmenu_command[5]="3 goto_menu"
set mainmenu_keycode[5]=112
set mainansi_caption[5]="Alternate ^[1mP^[mlatform Images..."

\ Will be reset by init_pi (see below and in menu-commands.4th).
set pimenu_optionstext="Platform Image: (UNINIT)"
\ For feeding init_pi.
set pitext="Platform Image: "

set menuset_name3="pi1"
set pi1menu_init[1]="init_pi"

set pi1menu_caption[1]="Back to Main Menu [Backspace]"
set pi1menu_command[1]="pi_draw_screen drop 1 goto_menu"
set pi1menu_keycode[1]=8
set pi1ansi_caption[1]="Back to Main Menu ^[1m[Backspace]^[m"

set pi1menu_caption[2]="[P]age: 1 of 2"
set pi1menu_command[2]="4 goto_menu"
set pi1menu_keycode[2]=112
set pi1ansi_caption[2]="^[1mP^[mage: 1 of 2"

set pi1menu_options=3
set pi1menu_optionstext="${pimenu_optionstext}"

set pi1menu_caption[3]="[D]efault (20210804T003855Z)"
set pi1menu_command[3]="s\" set bootpi=default\" evaluate pi_unload pi_draw_screen"
set pi1menu_keycode[3]=100
set pi1ansi_caption[3]="^[1mD^[mefault (20210804T003855Z)"


set pi1menu_caption[4]="20210729T002724Z"
set pi1menu_command[4]="pi_unload s\" set bootpi=20210729T002724Z\" evaluate s\" load /os/20210729T002724Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210729T002724Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[4]="20210729T002724Z"


set pi1menu_caption[5]="20210715T010227Z"
set pi1menu_command[5]="pi_unload s\" set bootpi=20210715T010227Z\" evaluate s\" load /os/20210715T010227Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210715T010227Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[5]="20210715T010227Z"


set pi1menu_caption[6]="20210701T204427Z"
set pi1menu_command[6]="pi_unload s\" set bootpi=20210701T204427Z\" evaluate s\" load /os/20210701T204427Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210701T204427Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[6]="20210701T204427Z"


set pi1menu_caption[7]="20210617T001230Z"
set pi1menu_command[7]="pi_unload s\" set bootpi=20210617T001230Z\" evaluate s\" load /os/20210617T001230Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210617T001230Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[7]="20210617T001230Z"


set pi1menu_caption[8]="20210520T001536Z"
set pi1menu_command[8]="pi_unload s\" set bootpi=20210520T001536Z\" evaluate s\" load /os/20210520T001536Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210520T001536Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[8]="20210520T001536Z"

set menuset_name4="pi2"
set pi2menu_init[1]="init_pi"

set pi2menu_caption[1]="Back to Main Menu [Backspace]"
set pi2menu_command[1]="pi_draw_screen drop 1 goto_menu"
set pi2menu_keycode[1]=8
set pi2ansi_caption[1]="Back to Main Menu ^[1m[Backspace]^[m"

set pi2menu_caption[2]="[P]age: 2 of 2"
set pi2menu_command[2]="3 goto_menu"
set pi2menu_keycode[2]=112
set pi2ansi_caption[2]="^[1mP^[mage: 2 of 2"

set pi2menu_options=3
set pi2menu_optionstext="${pimenu_optionstext}"

set pi2menu_caption[3]="[D]efault (20210804T003855Z)"
set pi2menu_command[3]="s\" set bootpi=default\" evaluate pi_unload pi_draw_screen"
set pi2menu_keycode[3]=100
set pi2ansi_caption[3]="^[1mD^[mefault (20210804T003855Z)"


set pi2menu_caption[4]="20210422T002312Z"
set pi2menu_command[4]="pi_unload s\" set bootpi=20210422T002312Z\" evaluate s\" load /os/20210422T002312Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210422T002312Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi2ansi_caption[4]="20210422T002312Z"

```

**Comment by danmcd** [comment]
Testing notes in ticket.

File: src/piadm.sh [hunk 1]
```diff
@@ -588,6 +593,151 @@ update_boot_sectors() {
...
+
+		pagenum=$((((pinum) / pisperpage) + 1))
+		itemnum=$(((pinum % pisperpage) + 1))
+		if [[ "$itemnum" == "1" ]]; then
+			menusetprefix=$(emit_pageheader $pagenum $totalpages \
+				$defaultpi)
+		fi
+		# else menusetprefix is already set for this round!
+
+		# Emit the entry.  XXX KEBE SAYS FILL ME IN!
```
Comment (papertigers) [inline src/piadm.sh:729]: Left over comment?

File: src/piadm.sh [hunk 2]
```diff
@@ -239,6 +239,7 @@ install() {
 		# URL_PREFIX.  Grab the latest-version ISO. Before proceeding,
 		# make sure it's the current one.
 		iso=yes
+		vecho "Downloading latest SmartOS ISO"
 		"${CURL[@]}" -o "${tdir}/smartos.iso" "${URL_PREFIX}/smartos-latest.iso"
```
Comment (bahamat) [inline src/piadm.sh:243]: Since we're being verbose now, consider adding `--progress-bar` to curl

File: src/piadm.sh [hunk 3]
```diff
@@ -293,6 +296,7 @@ install() {
 		# Explicit boot stamp or URL.
 
 		# Do a URL reality check.
+		vecho "Attempting download of $1"
 		"${CURL[@]}" -o "${tdir}/download" "$1"
```
Comment (bahamat) [inline src/piadm.sh:300]: Same with `--progress-bar` here

File: src/piadm.sh (hunk 1 above)
Comment (danmcd) [reply src/piadm.sh:729]: Will fix.

File: src/piadm.sh (hunk 2 above)
Comment (danmcd) [reply src/piadm.sh:243]: See below...

File: src/piadm.sh (hunk 3 above)
Comment (danmcd) [reply src/piadm.sh:300]: Let me see what I can do.  It's easy enough to test...

Comment (danmcd) [reply src/piadm.sh:300]: ... seems to work. Please look over the new diff, @bahamat !

**Review by bahamat — APPROVED** [review]
LGTM!

**Review by papertigers — APPROVED** [review]

**Pull request merged**